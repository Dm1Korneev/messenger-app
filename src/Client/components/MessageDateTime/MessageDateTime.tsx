import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import { PropsWithChildren } from 'react';

function getLang() {
  if (navigator.languages) {
    return navigator.languages[0];
  }
  return navigator.language;
}

const optionsLong: Intl.DateTimeFormatOptions = {
  weekday: 'long',
  month: 'long',
  day: 'numeric',
  year: 'numeric',
  hour: 'numeric',
  minute: '2-digit',
};
const dateLongFormat = new Intl.DateTimeFormat(getLang(), optionsLong);

const optionsShort: Intl.DateTimeFormatOptions = {
  hour: 'numeric',
  minute: '2-digit',
};
const dateShortFormat = new Intl.DateTimeFormat(getLang(), optionsShort);

const isToday = (someDate: Date) => {
  const today = new Date();
  return someDate.getDate() === today.getDate()
    && someDate.getMonth() === today.getMonth()
    && someDate.getFullYear() === today.getFullYear();
};

const getDateTimeText = (value: Date) => {
  if (value) {
    if (isToday(value)) {
      return dateShortFormat.format(value);
    }
    return dateLongFormat.format(value);
  }
  return undefined;
};

type MessageDateTimeProps = {
  isCurrentUserMessage: boolean;
  dateTime: Date;
}

export const MessageDateTime = ({
  isCurrentUserMessage, children, dateTime: dateTimeFromProps,
}: PropsWithChildren<MessageDateTimeProps>) => {
  const flexDirection = isCurrentUserMessage ? 'row' : 'row-reverse';
  const alignSelf = isCurrentUserMessage ? 'flex-start' : 'flex-end';

  return (
    <Box display="flex" flexDirection="column">
      <Box
        alignSelf={alignSelf}
        bgcolor="white"
        display="flex"
        flexDirection={flexDirection}
        position="sticky"
        top={0}
        width="50%"
        zIndex={1000}
      >
        <Typography
          color="textSecondary"
          variant="caption"
        >
          {getDateTimeText(dateTimeFromProps)}
        </Typography>
      </Box>
      {children}
    </Box>
  );
};
