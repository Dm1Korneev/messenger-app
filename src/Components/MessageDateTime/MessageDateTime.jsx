import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';

function getLang() {
  if (navigator.languages) {
    return navigator.languages[0];
  }
  return navigator.language;
}

const optionsLong = {
  weekday: 'long',
  month: 'long',
  day: 'numeric',
  year: 'numeric',
  dateStyle: 'full',
  timeStyle: 'short',
  minute: '2-digit',
};
const dateLongFormat = new Intl.DateTimeFormat(getLang(), optionsLong);

const optionsShort = {
  timeStyle: 'short',
  minute: '2-digit',
};
const dateShortFormat = new Intl.DateTimeFormat(getLang(), optionsShort);

const isToday = (someDate) => {
  const today = new Date();
  return someDate.getDate() === today.getDate()
    && someDate.getMonth() === today.getMonth()
    && someDate.getFullYear() === today.getFullYear();
};

const getDateTimeText = (value) => {
  if (value) {
    if (isToday(value)) {
      return dateShortFormat.format(value);
    }
    return dateLongFormat.format(value);
  }
  return undefined;
};

const useStyles = makeStyles((theme) => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 1000,
  },
}));

const MessageDateTime = ({
  isCurrentUserMessage, children, dateTime: dateTimeFromProps,
}) => {
  const classes = useStyles();

  const flexDirection = isCurrentUserMessage ? 'row' : 'row-reverse';
  const alignSelf = isCurrentUserMessage ? 'flex-start' : 'flex-end';

  return (
    <Box display="flex" flexDirection="column">
      <Box
        position="sticky"
        top={0}
        display="flex"
        width="50%"
        flexDirection={flexDirection}
        alignSelf={alignSelf}
        className={classes.ListItemText}
      >
        <Typography
          color="textSecondary"
          variant="caption"
        >
          {getDateTimeText(dateTimeFromProps)}
        </Typography>
      </Box>
      {children}
    </Box>
  );
};

MessageDateTime.propTypes = {
  isCurrentUserMessage: PropTypes.bool.isRequired,
  children: PropTypes.element.isRequired,
  dateTime: PropTypes.instanceOf(Object).isRequired,
};

export default MessageDateTime;
