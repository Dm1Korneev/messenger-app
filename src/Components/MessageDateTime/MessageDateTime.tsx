import React, { FC } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';

function getLang() {
  if (navigator.languages) {
    return navigator.languages[0];
  }
  return navigator.language;
}

const optionsLong: Intl.DateTimeFormatOptions = {
  weekday: 'long',
  month: 'long',
  day: 'numeric',
  year: 'numeric',
  hour: 'numeric',
  minute: '2-digit',
};
const dateLongFormat = new Intl.DateTimeFormat(getLang(), optionsLong);

const optionsShort: Intl.DateTimeFormatOptions = {
  hour: 'numeric',
  minute: '2-digit',
};
const dateShortFormat = new Intl.DateTimeFormat(getLang(), optionsShort);

const isToday = (someDate: Date) => {
  const today = new Date();
  return someDate.getDate() === today.getDate()
    && someDate.getMonth() === today.getMonth()
    && someDate.getFullYear() === today.getFullYear();
};

const getDateTimeText = (value: Date) => {
  if (value) {
    if (isToday(value)) {
      return dateShortFormat.format(value);
    }
    return dateLongFormat.format(value);
  }
  return undefined;
};

const useStyles = makeStyles((theme) => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 1000,
  },
}));

type Props = {
  isCurrentUserMessage: boolean;
  dateTime: Date;
}

export const MessageDateTime: FC<Props> = ({
  isCurrentUserMessage, children, dateTime: dateTimeFromProps,
}) => {
  const classes = useStyles();

  const flexDirection = isCurrentUserMessage ? 'row' : 'row-reverse';
  const alignSelf = isCurrentUserMessage ? 'flex-start' : 'flex-end';

  return (
    <Box display="flex" flexDirection="column">
      <Box
        position="sticky"
        top={0}
        display="flex"
        width="50%"
        flexDirection={flexDirection}
        alignSelf={alignSelf}
        className={classes.ListItemText}
      >
        <Typography
          color="textSecondary"
          variant="caption"
        >
          {getDateTimeText(dateTimeFromProps)}
        </Typography>
      </Box>
      {children}
    </Box>
  );
};
