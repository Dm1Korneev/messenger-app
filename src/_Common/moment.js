import moment from 'moment';
import 'moment/locale/ru';

function getLang() {
  if (navigator.languages) {
    return navigator.languages[0];
  }
  return navigator.language;
}

moment.locale(getLang());
export default moment;
