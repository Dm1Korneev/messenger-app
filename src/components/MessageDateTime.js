import React from "react";
import classNames from "classnames";
import PropTypes from "prop-types";
import moment from "../common/moment";

// @material-ui
import withStyles from "@material-ui/core/styles/withStyles";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Typography from "@material-ui/core/Typography";

function getDateTimeText(_value) {
  if (_value) {
    const value = moment(_value);
    if (value.isAfter(moment().startOf("day"))) {
      return value.format("LT");
    } else {
      return value.format("LLLL");
    }
  }
}

function MessageDateTime(props) {
  const { isCurrentUserMessage, children, classes } = props;
  const dateTime = getDateTimeText(props.dateTime);

  let ListItemTextClass = classNames(
    classes.ListItemText,
    !isCurrentUserMessage && classes.reverse
  );

  const ListItemClass = classNames(
    classes.ListItem,
    isCurrentUserMessage && classes.reverse
  );

  return (
    <ListItem className={ListItemClass}>
      <div className={classes.textContent}>
        <ListItemText
          className={classNames(ListItemTextClass, classes.sticky)}
          secondary={
            <Typography
              component="span"
              color="textSecondary"
              className={classes.ListItemText__dateTime}
            >
              {dateTime}
            </Typography>
          }
        />
        {children}
      </div>
    </ListItem>
  );
}

MessageDateTime.propTypes = {
  classes: PropTypes.object.isRequired,
  isCurrentUserMessage: PropTypes.bool.isRequired,
  children: PropTypes.element
};

const styles = theme => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 1000,
    width: "50%",
    display: "flex",
    justifyContent: "space-between",
    "&:first-child": {
      paddingLeft: theme.spacing.unit * 2
    }
  },
  reverse: {
    flexDirection: "row-reverse",
    alignSelf: "flex-end"
  },
  ListItem: {
    alignItems: "start",
    paddingTop: 0,
    paddingBottom: 0,
    paddingRight: 0,
    paddingLeft: 0
  },
  textContent: {
    flexGrow: 1,
    display: "flex",
    flexDirection: "column"
  },
  ListItemText__dateTime: {
    fontSize: theme.typography.pxToRem(11)
  },
  sticky: {
    position: "sticky",
    top: 0
  }
});

export default withStyles(styles)(MessageDateTime);
