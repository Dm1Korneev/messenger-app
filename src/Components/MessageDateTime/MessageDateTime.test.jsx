import React from 'react';

import moment from 'Common/moment';
import MessageDateTime from './MessageDateTime';

jest.mock('@material-ui/core/ListItem', () => global.mockComponent('ListItem'));
jest.mock('@material-ui/core/ListItemText', () => global.mockComponent('ListItemText'));
jest.mock('@material-ui/core/Typography', () => global.mockComponent('Typography'));

const props = {
  isCurrentUserMessage: false,
  children: <></>,
  dateTime: moment.utc('20010101'),
};

describe('render MessageDateTime component', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = global.mount(
      <MessageDateTime {...props} />,
    );
  });

  test('component is render', () => {
    expect(wrapper.find('MessageDateTime').length).toBe(1);
  });

  test('ListItem subcomponent is render', () => {
    expect(wrapper.find({ originalcomponent: 'ListItem' }).length).toBe(1);
  });
});
