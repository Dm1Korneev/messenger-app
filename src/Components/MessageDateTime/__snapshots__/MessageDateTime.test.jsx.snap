// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MessageDateTime component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="ListItem"
    test-prop-classname="\\"MessageDateTime-ListItem-3\\""
  >
    <div
      class="MessageDateTime-textContent-4"
    >
      <div
        data-originalcomponent="ListItemText"
        test-prop-classname="\\"MessageDateTime-ListItemText-1 MessageDateTime-reverse-2 MessageDateTime-sticky-6\\""
        test-prop-secondary="<div data-originalcomponent=\\"Typography\\" test-prop-component=\\"&quot;span&quot;\\" test-prop-color=\\"&quot;textSecondary&quot;\\" test-prop-classname=\\"&quot;MessageDateTime-ListItemText__dateTime-5&quot;\\" data-reactroot=\\"\\">Monday, January 1, 2001 12:00 AM</div>"
      />
    </div>
  </div>
</DocumentFragment>
`;
